Makefile.common

```diff
@@ -10,176 +10,147 @@
-TOPBYTE:=$(TOPBINOPT:.opt$(EXE)=.byte$(EXE))
-
-COQTOPEXE:=bin/coqtop$(EXE)
-COQTOPBYTE:=bin/coqtop.byte$(EXE)
-
-COQDEP:=bin/coqdep$(EXE)
-COQPP:=bin/coqpp$(EXE)
-DOC_GRAM:=bin/doc_grammar$(EXE)
-COQDEPBYTE:=bin/coqdep.byte$(EXE)
-COQMAKEFILE:=bin/coq_makefile$(EXE)
-COQMAKEFILEBYTE:=bin/coq_makefile.byte$(EXE)
-COQTEX:=bin/coq-tex$(EXE)
-COQTEXBYTE:=bin/coq-tex.byte$(EXE)
-COQWC:=bin/coqwc$(EXE)
-COQWCBYTE:=bin/coqwc.byte$(EXE)
-COQDOC:=bin/coqdoc$(EXE)
-COQDOCBYTE:=bin/coqdoc.byte$(EXE)
-COQC:=bin/coqc$(EXE)
-COQCOPT:=bin/coqc.opt$(EXE)
-COQCBYTE:=bin/coqc.byte$(EXE)
-COQWORKMGR:=bin/coqworkmgr$(EXE)
-COQWORKMGRBYTE:=bin/coqworkmgr.byte$(EXE)
+BCONTEXT=_build/default
+CONTEXT=_build/install/default
+CBIN=_build/install/default/bin
+CSHARE=_build/install/default/share/coq
+
+ifeq ($(BEST),byte)
+  COQC:=$(CBIN)/coqc.byte$(EXE)
+  export CAML_LD_LIBRARY_PATH:=$(shell echo $(CONTEXT)/lib/stublibs:$$CAML_LD_LIBRARY_PATH)
+else
+  COQC:=$(CBIN)/coqc$(EXE)
+endif
+
+COQTOPEXE:=$(CBIN)/coqtop$(EXE)
+TOPBINOPT:=$(addsuffix .opt$(EXE), $(addprefix $(CBIN)/, coqproofworker coqtacticworker coqqueryworker)) $(COQTOPEXE)
+COQDEP:=$(CBIN)/coqdep$(EXE)
+DOC_GRAM:=_build/default/doc/tools/docgram/doc_grammar.exe
+COQMAKEFILE:=$(CBIN)/coq_makefile$(EXE)
+COQTEX:=$(CBIN)/coq-tex$(EXE)
+COQWC:=$(CBIN)/coqwc$(EXE)
+COQDOC:=$(CBIN)/coqdoc$(EXE)
+COQDOCSTY:=$(CONTEXT)/lib/coq-core/tools/coqdoc/coqdoc.sty
+COQDOCCSS:=$(CONTEXT)/lib/coq-core/tools/coqdoc/coqdoc.css
+COQWORKMGR:=$(CBIN)/coqworkmgr$(EXE)
 COQMAKE_ONE_TIME_FILE:=tools/make-one-time-file.py
 COQTIME_FILE_MAKER:=tools/TimeFileMaker.py
 COQMAKE_BOTH_TIME_FILES:=tools/make-both-time-files.py
 COQMAKE_BOTH_SINGLE_TIMING_FILES:=tools/make-both-single-timing-files.py
-VOTOUR:=bin/votour
-OCAMLLIBDEP:=bin/ocamllibdep$(EXE)
-OCAMLLIBDEPBYTE:=bin/ocamllibdep.byte$(EXE)
+VOTOUR:=$(CBIN)/votour
+OCAMLLIBDEP:=$(CBIN)/ocamllibdep$(EXE)
+FAKEIDE:=$(CBIN)/fake_ide$(EXE)
+USERCONTRIBDIRS:=Ltac2
+CHICKEN:=$(CBIN)/coqchk
+TOOLS:=$(VOTOUR) $(COQDOC) $(COQWC)
+CSDPCERT:=$(CBIN)/csdpcert
+
+ifeq ($(origin COQ_SRC_DIR),undefined)
+COQ_SRC_DIR=.
+endif
 
-# these get installed!
-TOOLS:=$(COQDEP) $(COQMAKEFILE) $(COQTEX) $(COQWC) $(COQDOC) $(COQC)\
-	$(COQWORKMGR) $(COQPP) $(VOTOUR) $(OCAMLLIBDEP)
-TOOLS_HELPERS:=tools/CoqMakefile.in $(COQMAKE_ONE_TIME_FILE) $(COQTIME_FILE_MAKER)\
-	$(COQMAKE_BOTH_TIME_FILES) $(COQMAKE_BOTH_SINGLE_TIMING_FILES)
+COQ_CM_LIBS=coqpp lib clib kernel library engine pretyping gramlib interp printing parsing proofs tactics vernac stm toplevel topbin
+ML_SOURCE_DIRS=$(addprefix $(COQ_SRC_DIR)/,$(COQ_CM_LIBS))
+ALL_ML_SOURCE_FILES=$(shell find $(ML_SOURCE_DIRS) -name '*.ml' -or -name '*.mli' -or -name '*.c' -or -name '*.h')
 
-COQDEPBOOT:=bin/coqdep_boot$(EXE)
-COQDEPBOOTBYTE:=bin/coqdep_boot.byte$(EXE)
-FAKEIDE:=bin/fake_ide$(EXE)
-FAKEIDEBYTE:=bin/fake_ide.byte$(EXE)
+# Override for developer build [to get warn-as-error for example]
+_DPROFILE?=release
+_DOPT:=--display quiet --profile=$(_DPROFILE)
```

Comment:
As already mentioned, please remove this definition of `_DOPT`. Calling `make` should respect the default behavior of `dune`. It is not the job of `make` to pass extra options, unless the user explicitly asked for it.

Another option is to set `_DPROFILE ?= dev`, which I believe is the default behavior of `dune`, so that the user can type `make _DPROFILE=release` to override it. (But since there is already an environment variable `DUNE_PROFILE` which plays this role, it is better to just remove `_DPROFILE`.)

Reply:
Sorry my comment got lost, `--profile=dev` is needed as to preserve the semantics of the make-based build system, which builds binaries without opaque and readily installable. Otherwise, calling `make && make install` would produce a build not suitable for install. Feel free to override `_DOPT` locally if you need it, otherwise just switch to the dune build if you need dune features.

Reply:
That's the reason `_DPROFILE` is guareded with a `?` mark

Reply:
Since when is this pull request meant to preserve the semantics of `make && make install`? Did you even try it? Because it definitely does not work for me!

Reply:
What problem you are facing? Indeed I have tried it and as you can see the CI passed a few times, which indeed uses `make` and `make install`

Reply:
Anyways what you are asking is just _wrong_ , --profile cannot be set to dev if you expect the resulting files to be installable.

Reply:
```console
$ git clean -fxd
$ ./configure -prefix /tmp/coq-dune
$ make && make install
...
Installing /tmp/coq-dune/doc/coqide-server/README.md
INSTALL   VOFILES
install: cannot stat '_build_vo/default/theories/Init/Tauto.vo': No such file or directory
...
install: cannot stat '_build_vo/default/user-contrib/Ltac2/Notations.vo': No such file or directory
make[1]: *** [Makefile.install:74: install-library] Error 1
```

Reply:
Ups, for some reason the defaul target for the makefile is botched, wasn't caught in the CI as it always uses `make world` , let me have a look... makefile is such a PITA.

Reply:
Ok, problem fixed, that took me 1 hour.... The problem was that the current `submake` infrastructure is kinda weird (including master), using `MAKECMDGOALS` means that the default goals of the main make will be just ignored and instead the sumakefile will be called with the default goal (empty).

All this could be cleaned up a lot I think, but it is beyond my abilities or interest [still #13890 is a good start]

Reply:
@silene Are you OK with the explanations and the fix? I.e., that by default `make` will preserve the "installable" semantics and that one should prefer `make -f Makefile.dune` or call `dune` directly for actual development.

Reply:
No, I am not. For example, if I pass `-local` or `-profile devel` to `./configure`, why would `make` not behave the same as `dune`? I mean, there is "Nothing to install in a local build!", as `make install` rightfully complains (even after this pull request).

Reply:
Right. @ejgallego It seems that the Dune profile should really depend on the `configure` profile. Either, you build in release mode because `-prefix` was used or you build in dev mode because `-local` was used. This would solve the issue of keeping two incompatible systems.

Reply:
Or you can remove `-local` as long as you ensure `-profile devel` will use the good Dune mode for developers.